apis/placement/v1beta1/clusterresourceplacement_types.go

```diff
@@ -225,11 +225,142 @@ type PreferredClusterSelector struct {
+	// of a given metric is greater than or equal to the value specified in the requirement.
+	MetricSelectorGreaterThanOrEqualTo MetricSelectorOperator = "Ge"
+	// MetricSelectorEqualTo dictates Fleet to select cluster if its observed value of a given
+	// metric is equal to the values specified in the requirement.
+	MetricSelectorEqualTo MetricSelectorOperator = "Eq"
+	// MetricSelectorNotEqualTo dictates Fleet to select cluster if its observed value of a given
+	// metric is not equal to the values specified in the requirement.
+	MetricSelectorNotEqualTo MetricSelectorOperator = "Ne"
+	// MetricSelectorLessThan dictates Fleet to select cluster if its observed value of a given
+	// metric is less than the value specified in the requirement.
+	MetricSelectorLessThan MetricSelectorOperator = "Lt"
+	// MetricSelectorLessThanOrEqualTo dictates Fleet to select cluster if its observed value of a
+	// given metric is less than or equal to the value specified in the requirement.
+	MetricSelectorLessThanOrEqualTo MetricSelectorOperator = "Le"
+)
+
+// MetricSelectorRequirement is a specific metric requirement when picking clusters for
+// resource placement.
+type MetricSelectorRequirement struct {
+	// Name is the name of the metric; it should be a Kubernetes label name.
+	// +required
+	Name string `json:"name"`
+
+	// Operator specifies the relationship between a cluster's observed value of the specified
+	// metric and the values given in the requirement.
+	// +required
+	Operator MetricSelectorOperator `json:"operator"`
+
+	// Values are a list of values of the specified metric which Fleet will compare against
+	// the observed values of individual member clusters in accordance with the given
+	// operator.
+	//
+	// If the operator is Gt (greater than), Ge (greater than or equal to), Lt (less than),
+	// or `Le` (less than or equal to), exactly one value must be specified in the list.
+	//
+	// If the operator is Eq (equal to), or Ne (ne), at least one value must be specified
+	// in the list.
 	// +required
+	Values []string `json:"values"`
+}
+
+// MetricSelector helps user specify metric requirements when picking clusters for resource
+// placement.
+type MetricSelector struct {
+	// MatchExpressions is an array of MetricSelectorRequirements. The requirements are AND'd.
+	// +optional
+	MatchExpressions []MetricSelectorRequirement `json:"matchMetrics"`
+}
+
+type MetricSortPreference struct {
+	// Name is the name of the metric which Fleet sorts clusters by.
+	Name string `json:"name"`
+
+	// Prefer explains how Fleet should perform the sort; specifically, whether Fleet should
+	// sort in ascending or descending order.
+	Prefer MetricSortOrder `json:"prefer"`
+}
+
+type MetricSorter struct {
+	// SortByMetric is an array of MetricSortPreference. At most one sort preference can be
+	// specified in each term.
+	//
+	// +kubebuilder:validation:MaxItems:1
+	// +optional
+	SortByMetric []MetricSortPreference `json:"sortByMetric"`
```

Comment:
so this is an array of 1 now? How can we handle more than 1 in the future?

Reply:
Shall we only support one instead? Customer can always use N PreferredClusterSelector each containing a different metric and a different weight.

Then this becomes:

```
type MetricSorter struct {
  Name string
  Order MetricSortOrder
}
```

Reply:
Hi Liqian and Ryan! Originally the field is marked as an array with a max. length of 1 to avoid confusing default values. The parent field has been updated to use a pointer to achieve the same result, so this array is no longer needed.